src/type_url.rs

```diff
@@ -0,0 +1,19 @@
+//! Support for associating a type URL with a [`Message`].
+
+use crate::Message;
+
+/// Associate a type URL with the given [`Message]`.
+pub trait TypeUrl: Message {
+    /// Type URL for this [`Message`]. They take the form:
+    ///
+    /// ```text
+    /// /<package>.<TypeName>
+    /// ```
+    ///
+    /// For example:
+    ///
+    /// ```text
+    /// /foo.bar.baz.MyTypeName
```

Comment:
I don't think the slash is part of the type URL, the type name is `foo.bar.baz.MyTypeName` which then gets encoded into the Any message as `type.googleapis.com/foo.bar.baz.MyTypeName`

Reply:
Hmm, I guess this particular encoding is an [artifact of the type URLs I'm working with](https://github.com/cosmos/cosmos-sdk/issues/10949) and not the standard encoding.

I'm still unclear though if all type URLs are expected to be prefixed with `https://type.googleapis.com/` or just Google's own well-known types, which is what the example is for.

Reply:
There is difference between **type url** and **full type name** of a protobuf message.

For instance the **full type name** is not expected to have slashes.

The **Type URL** in a `google.Protobuf.Any`  is expected to [have at least one slash](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/any.proto#L130).

So if this Type URL mentioned here is supposed to end up in an any then by specification, it should contain at least one slash.

Reply:
If this **TYPE URL** we are talking about in this PR is supposed to be used to compose the TYPE URL of an any then we shouldn't be talking about *TYPE URL* but we should be talking about **FULLNAME* of a protobuf  message, they're different concepts.

Reply:
Yes, the type url in Any should have a slash, but the type name that we attach to the trait should just be the full type name without a slash.

Reply:
Then this trait should not be called `TypeURL`  it should be called `Fullname`. 

And the way we match a type URL is not the correct one:
- Not compliant: https://github.com/tokio-rs/prost/pull/858/files#diff-fe4eba36fffb306e25d27bf29ddea9cd1634ea4ad86705045c1cb13a3aa27346R59

- Compliant(golang): https://github.com/protocolbuffers/protobuf-go/blob/v1.31.0/types/known/anypb/any.pb.go#L337

Reply:
The use case is populating the `type_url` field of an `Any` message.

If you're saying the type URL can always be computed from the fullname by prepending a slash, that's fine. That's what the implementation I'm using is expecting.